src/utils/subgraph.ts

```diff
@@ -0,0 +1,120 @@
+import { gql, DocumentNode, ApolloQueryResult } from '@apollo/client'
+import { DDO } from '@oceanprotocol/lib'
+import { getApolloClientInstance } from '../providers/ApolloClientProvider'
+import BigNumber from 'bignumber.js'
+
+const freQuery = gql`
+  query AssetFrePrice($datatoken_in: [String!]) {
+    fixedRateExchanges(orderBy: id, where: { datatoken_in: $datatoken_in }) {
+      rate
+      id
+      datatoken {
+        id
+        address
+      }
+    }
+  }
+`
+
+const poolQuery = gql`
+  query AssetPoolPrice($datatokenAddress_in: [String!]) {
+    pools(where: { datatokenAddress_in: $datatokenAddress_in }) {
+      spotPrice
+      id
+      datatokenAddress
+    }
+  }
+`
+
+const previousOrderQuery = gql`
+  query AssetPreviousOrder($id: String!, $account: String!) {
+    tokenOrders(
+      first: 1
+      where: { datatokenId: $id, payer: $account }
+      orderBy: timestamp
+      orderDirection: desc
+    ) {
+      timestamp
+      tx
+    }
+  }
+`
+
+async function fetchData(
+  query: DocumentNode,
+  variables: any
+): Promise<ApolloQueryResult<any>> {
+  try {
+    const client = getApolloClientInstance()
+    const response = await client.query({
+      query: query,
+      variables: variables
+    })
+    return response
+  } catch (error) {
+    console.error('Error fetchData: ', error.message)
+  }
+}
+
+export async function getPreviousOrders(
+  id: string,
+  account: string,
+  assetTimeout: string
+): Promise<string> {
+  const variables = {
+    id: id,
+    account: account
+  }
+  const fatchedPreviousOrders: any = await fetchData(
+    previousOrderQuery,
+    variables
+  )
+  if (fatchedPreviousOrders.data?.tokenOrders?.length === 0) return null
+  if (assetTimeout === 'Forever') {
+    return fatchedPreviousOrders?.data?.tokenOrders[0]?.tx
+  } else {
+    const expiry = new BigNumber(
+      fatchedPreviousOrders?.data?.tokenOrders[0]?.timestamp
+    ).plus(assetTimeout)
+    const unixTime = new BigNumber(Math.floor(Date.now() / 1000))
+    if (unixTime.isLessThan(expiry)) {
+      return fatchedPreviousOrders?.data?.tokenOrders[0]?.tx
+    } else {
+      return null
+    }
+  }
+}
+
+export async function getAssetPrice(assets: DDO[]): Promise<any> {
```

Comment:
this is nice, and I can see how this could replace relying on `ddo.price.value` in other places, like on asset teasers in all search query lists.

So we should be precise and also type the response, not just `any`. Would call method plural then, `getAssetPrices()`

Reply:
Updating method name now, but not sure how can we be more strict with return type because it's an object so it helps us accessing the price using the dataset did in populating asset selection list without iterating the returned price list again
like this :  price: priceList[did]

Reply:
yup, that's all fine, what I meant is that we know the structure of the object returned, so instead of typing it as `any` we should be precise. If I quickly see this right, we are also returning `DDO[]`, right? So should be typed as `Promise<DDO[]>` then

Reply:
no, actually we return a object not an array, eg object =  { 'didAsset1': priceValueAsset1, 'didAsset2': priceValueAsset2}

Reply:
so then sounds like the typing should be `{ [key: string]: number }` or `{ [key: string]: string }` instead of `any`? We repeat that `any` then in many places so would make sense creating interface `PriceList` for it here and import that wherever we need it. But if we correctly type the return here there will be no need to do stuff like this in other places:

```ts
const priceList: any = await getAssetPrices(ddoList)
```

Cause then the typing is inferred from usage so this will be enough instead:

```ts
const priceList = await getAssetPrices(ddoList)
```

Reply:
yeah that could work, having an interface PriceList : { [key: string]: string } an using it